utilities/ggd/lib/dag.go

```diff
@@ -0,0 +1,154 @@
+	"fmt"
+	"go/build"
+
+	"golang.org/x/tools/go/buildutil"
+)
+
+// ImportSet is a set of imports.
+type ImportSet map[string]struct{}
+
+// ImportGraph is a map from a Go package, `p` to the set of all packages which import `p`.
+type ImportGraph map[string]ImportSet
+
+// String representation of the Import Graph.
+func (g ImportGraph) String() string {
+	var buffer bytes.Buffer
+	for pkg, importers := range g {
+		for importer := range importers {
+			buffer.WriteString(fmt.Sprintf("%s imported by %s\n", pkg, importer))
+		}
+	}
+	return buffer.String()
+}
+
+func NewImportGraph(ctx *build.Context, basePkg string) (ImportGraph, error) {
+	pkgs := buildutil.ExpandPatterns(ctx, []string{basePkg + "/..."})
+	g := make(ImportGraph, len(pkgs))
+
+	for pkg := range pkgs {
+		buildPkg, err := ctx.Import(pkg, "", 0)
+		if err != nil {
+			if _, ok := err.(*build.NoGoError); ok {
+				// The package doesn't contain any buildable go files. This can be caused, for
+				// example, when the only go files in a package contain build flags which are
+				// not present in our build context.
+				continue
+			}
+			return nil, fmt.Errorf("could not import package %v: %v", pkg, err)
+		}
+
+		visited := make(map[string]string)
+
+		// ensure we add the pkg to the graph
+		if _, ok := g[pkg]; !ok {
+			g[pkg] = make(ImportSet)
+		}
+
+		for _, path := range buildPkg.Imports {
+			g.maybeAddEdge(ctx, buildPkg, visited, pkg, path)
+		}
+		for _, path := range buildPkg.TestImports {
+			g.maybeAddEdge(ctx, buildPkg, visited, pkg, path)
+		}
+		for _, path := range buildPkg.XTestImports {
+			g.maybeAddEdge(ctx, buildPkg, visited, pkg, path)
+		}
+	}
+
+	return g, nil
+}
+
+func (g ImportGraph) maybeAddEdge(
+	ctx *build.Context, buildPkg *build.Package, visited map[string]string, pkg, path string,
+) {
+	if path == "C" {
+		// Not a real package.
+		return
+	}
+
+	importedPkg, ok := visited[path]
+	if !ok {
+		// It's okay for Import to return an error as not all packages that can be found in
+		// a package will necessarily be present. For example, packages imported only by test
+		// files in vendored packages will not be installed. In the case of an error, Import
+		// always returns a non-nil *Package that contains partial information.
+		importBuildPkg, _ := ctx.Import(path, buildPkg.Dir, build.FindOnly)
+		if importBuildPkg != nil {
+			importedPkg = importBuildPkg.ImportPath
+		} else {
+			importedPkg = path
+		}
+		visited[path] = importedPkg
+		g.addEdge(importedPkg, pkg)
+	}
+}
+
+func (g ImportGraph) addEdge(from, to string) {
+	edges, ok := g[from]
+	if !ok {
+		edges = make(map[string]struct{})
+		g[from] = edges
+	}
+	edges[to] = struct{}{}
+}
+
+// Closure return the transitive closure of all packages reachable
+// by starting at the provided paths in the ImportGraph.
+func (g ImportGraph) Closure(paths ...string) (ImportSet, error) {
+	// add all the paths we're starting at, as they are by definition
+	// reachable.
+	for _, p := range paths {
```

Comment:
Instead of including the logic to determine if a package was deleted so we can add all packages here would it make sense to instead only add the package if it exists in the graph since otherwise we know it was deleted?

Reply:
> "only add the package if it exists in the graph since otherwise we know it was deleted" 

I actually tried to do this the first. Unfortunately, it doesn't work in certain cases. 

Can repro yourself, steps:

```
$ cd $GOPATH/src/github.com/m3db/build-tools
$ git checkout prateek/ggd/demo-testdata-issue
$ cd ./utilities/ggd
$ ./test.sh
# testcase6 will fail
# can see output for the run in ./utilities/ggd/bin/testcase6.out
```

Reply:
talked offline: if we detect changes in `*/testdata/*`, run everything; otherwise - do ^^ some version of what Jerome's suggesting.

Reply:
Done. There's one more edge case we didn't talk about: a dangling package delete. i.e. a package that's not imported by any other packages, and gets deleted. It isn't present the import graph either. I throw an error for it and run tests on all non-filtered packages.